pkg/k2/constructs/dynamic_value.go

```diff
@@ -0,0 +1,325 @@
+	l := zap.L()
+
+	ownerType := reflect.TypeOf(data.currentOwner).Kind().String()
+	ownerString := "unknown"
+
+	l = l.With(zap.String(ownerType, ownerString))
+
+	switch strings.ToLower(level) {
+	case "debug":
+		l.Sugar().Debugf(message, args...)
+	case "info":
+		l.Sugar().Infof(message, args...)
+	case "warn":
+		l.Sugar().Warnf(message, args...)
+	case "error":
+		l.Sugar().Errorf(message, args...)
+	default:
+		l.Sugar().Warnf(message, args...)
+	}
+	return ""
+}
+
+// toJson is used to return complex values that do not have TextUnmarshaler implemented
+func (ctx DynamicValueContext) toJson(value any) (string, error) {
+	j, err := json.Marshal(value)
+	if err != nil {
+		return "", err
+	}
+	return string(j), nil
+}
+
+func (ctx DynamicValueContext) PathAncestor(path construct.PropertyPath, depth int) (string, error) {
+	if depth < 0 {
+		return "", fmt.Errorf("depth must be >= 0")
+	}
+	if depth == 0 {
+		return path.String(), nil
+	}
+	if len(path) <= depth {
+		return "", fmt.Errorf("depth %d is greater than path length %d", depth, len(path))
+	}
+	return path[:len(path)-depth].String(), nil
+}
+
+func (ctx DynamicValueContext) PathAncestorExists(path construct.PropertyPath, depth int) bool {
+	return len(path) > depth
+}
+
+// FieldRef returns a reference to `field` on `resource` (as a PropertyRef)
+func (ctx DynamicValueContext) FieldRef(field string, resource any) (construct.PropertyRef, error) {
+	resId, err := TemplateArgToRID(resource)
+	if err != nil {
+		return construct.PropertyRef{}, err
+	}
+
+	return construct.PropertyRef{
+		Resource: resId,
+		Property: field,
+	}, nil
+}
+
+func TemplateArgToRID(arg any) (construct.ResourceId, error) {
+	switch arg := arg.(type) {
+	case construct.ResourceId:
+		return arg, nil
+
+	case construct.Resource:
+		return arg.ID, nil
+
+	case string:
+		var resId construct.ResourceId
+		err := resId.UnmarshalText([]byte(arg))
+		return resId, err
+	}
+
+	return construct.ResourceId{}, fmt.Errorf("invalid argument type %T", arg)
+}
+
+type DynamicValueSelection struct {
+	Source  any
+	mapKeys []reflect.Value
+	next    int
+	Key     string
+	Value   any
+	Index   int
+}
+
+func SelectItem(src any) DynamicValueSelection {
+	srcValue := reflect.ValueOf(src)
+	switch srcValue.Kind() {
+	case reflect.Map:
+		if !srcValue.IsValid() || srcValue.Len() == 0 {
+			return DynamicValueSelection{
+				Source: src,
+			}
+		}
+		keys := srcValue.MapKeys()
+		slices.SortStableFunc(keys, func(i, j reflect.Value) int {
+			return strings.Compare(stringValue(i.Interface()), stringValue(j.Interface()))
+		})
+		if len(keys) == 0 {
+			return DynamicValueSelection{
+				Source: src,
+			}
+		}
+		return DynamicValueSelection{
+			Source:  src,
+			mapKeys: keys,
+		}
+	default:
+		return DynamicValueSelection{
+			Source: src,
+		}
+	}
+}
+
+// Next returns the next value in the selection and whether there are more values
+func (s *DynamicValueSelection) Next() (any, bool) {
```

Comment:
This function can't be used in templates, probably worth documenting it as such.

> The name of a niladic method of the data, preceded by a period, such as .Method The result is the value of invoking the method with dot as the receiver, dot.Method(). Such a method must have one return value (of any type) or two return values, the second of which is an error.

Reply:
You are correct. I'll update the documentation.